extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/interceptor/TransactionalInterceptorBase.java

```diff
@@ -119,6 +125,8 @@ protected Object invokeInOurTx(InvocationContext ic, TransactionManager tm, Runn
                 tm.setTransactionTimeout(currentTmTimeout);
             }
         }
+        // put the transaction configuration inside the synchronization registry to access it from Hibernate
+        transactionSynchronizationRegistry.putResource(CONFIG_KEY, configAnnotation);
```

Comment:
I think that we also need to make sure that the transaction is rolled back rather than committed if it is read only.

Reply:
I'm not sure about this one. Read only transactions are only a hint sent to the underlying data-remated technolohy (here, the Hibernate Session), they are not realy a property of the transaction. If you use transaction with JDBC (or MongoDB), they will have no effect (this needs to be documented as a hint not honored by all data access layer).

Of course, we can set the transaction to be rollback only at the transaction manager level. But read only transaction are not  a JTA standard so I will be reluctant that they act at JTA level.

Reply:
Two things to keep in mind:
 - some databases can actually start a "read only transaction", and this will get you performance benefits; but to take advantage of this, you need to be able to start the transaction in RO mode from the very beginning. Might be out of scope? e.g. I don't think Agroal allows this today.
 - some (other) databases actually perform better on commit than on rollback. So I'm not sure about rolling back being a good idea as an enforced mode - I'm aware it's popular among some people, but I'd not be sure about enforcing this.

Why would you roll it back? Just to be really sure that no changes are written? To achieve such an effect, I think it would be more useful to actually wrap our data-access APIs with a read-only decoration, so to throw a clear exception on any attempt.

If we wanted to do that, I suppose we'd need the TX to be marked as transaction since its creation (so to affect any resources we provide within its scope); which would have the added benefit of allowing optimisation #1 above, should we want to implement that in the future.

Reply:
I just don't like the idea that a read only TX can still commit data, and if this does only affect hibernate it just feels wrong that you could still use JDBC to do writes in a supposed read only TX.

Reply:
I agree with that. Would be nice to be able to wrap such other APIs to prevent it.

Reply:
@stuartwdouglas @Sanne we already discuss this, it's not really a read only transaction. It's a transaction hint that only read only requests should be performed. Currently this hint is only used by Hibernate, but who knows, maybe something else will be able to handle it later ?

Another solution could be to move this inside the Hibernate extension, create a `@SessionConfiguration` annotation and implement an hibernate only solution (with a new CDI interceptor). This would maybe be cleaner but harder to find for users (more annotation, more doc, ...). This could still re-use the synchronization registry to pass the configuration from the interceptor to the entity manager.

Users comming from Spring are used to it because it's what they already have, a readOnly attribute on the Spring `@Transactional` annotation ...

But I must agree that the fact a user can use raw JDBC queries and specifying a readOnly that does nothing can be misleading ...

Reply:
So we have two options I believe we all could agree on:

1. convert this into a `@SessionConfiguration` so that it's clearer that this is a tuning *hint* for Hibernate only.
2. Implement a *real* read-only transaction support.

While 2. is probably more valuable, it's clearly a long way to get there: at very least it needs patches in multiple components.

So let's go with approach 1 and make it very clear that:

- this is Hibernate specific
- it's only a hint

Hopefully such annotation would be used for several more useful tuning options? It begs the question of how a single annotation becomes extensible, maybe it will need multiple annotations?

Reply:
OK for converting to `@SessionConfiguration`. 
There is two ways of implementing this:

1. Use a `@Transactional` CDI interceptor that then try to locate this `@SessionConfiguration` annotation: the session configuration will have the same scope as the transaction, and we can still use the Transaction Synchronization Registry to communicate between the interceptor and the entity manager factory.

2. Use a `@SessionConfiguration` CDI interceptor: in this case we cannot be sure that we are in the context of a transaction so we cannot use the Transaction Synchronization Registry as a communication layer between the interceptor and the entity manager factory. We can still use Arc inside the interceptor to access the entity manager. If we implement this, a user could  have a read only scope smaller than the transaction one (and possibly switch to read only / read write multiple times inside the same transaction).

WDYT ?

Reply:
@Sanne can you give your advice on how to implement read only sessions with `@SessionConfiguration` as we discuss in this thread ?
I proposed two way to do this and need your input on which one is better ;)

Reply:
@Sanne can you give your advise on the best way to update this old PR to use `@SessionConfiguration` as we discussed ? I proposed two ways here https://github.com/quarkusio/quarkus/pull/10077#discussion_r442700574 can you give me some feedback ?

Reply:
I would go with a take on approach 1) (some of this may be slightly overkill)

- Add an @TransactionConfig meta annotation
- @SessionConfiguration is annotated with @TransactionConfig
- The transaction interceptors are modified to discover @TransactionConfig annotations on the method or class, and put them into the TSR
- Have hibernate pull this info out of the TSR and use it to decide on how to proceed

There will be caching needed here, we don't want to be reading annotations every transaction. I am not sure if @TransactionConfig is really needed, but the alternatives are hard coding a hibernate annotation into the TX subsystem, or sticking every annotation into the TSR (which may end up having perf impact). 

@Sanne does this sound reasonable?